Lib/unittest/mock.py

```diff
@@ -2136,29 +2136,45 @@ def __init__(self, /, *args, **kwargs):
         code_mock.co_flags = inspect.CO_COROUTINE
         self.__dict__['__code__'] = code_mock
 
-    async def _mock_call(self, /, *args, **kwargs):
-        try:
-            result = super()._mock_call(*args, **kwargs)
-        except (BaseException, StopIteration) as e:
-            side_effect = self.side_effect
-            if side_effect is not None and not callable(side_effect):
-                raise
-            return await _raise(e)
+    async def _execute_mock_call(self, /, *args, **kwargs):
+        # This is nearly just like super(), except for sepcial handling
+        # of coroutines
 
         _call = self.call_args
+        self.await_count += 1
+        self.await_args = _call
+        self.await_args_list.append(_call)
 
-        async def proxy():
-            try:
-                if inspect.isawaitable(result):
-                    return await result
-                else:
-                    return result
-            finally:
-                self.await_count += 1
-                self.await_args = _call
-                self.await_args_list.append(_call)
+        effect = self.side_effect
+        if effect is not None:
+            if _is_exception(effect):
+                raise effect
+            elif not _callable(effect):
+                try:
+                    result = next(effect)
+                except StopIteration:
+                    # It is impossible to propogate a StopIteration
+                    # through coroutines because of PEP 479
+                    raise StopAsyncIteration
+                if _is_exception(result):
+                    raise result
+            elif inspect.iscoroutinefunction(effect):
```

Comment:
Spoke with Jason offline about reasonings behind using `inspect.iscoroutinefunction` vs `asyncio.iscoroutinefunction` vs the `mock._is_async_func` function. Will leave it as `inspect` to discourage the usage of the @coroutines which are going away in 3.10 (that `asyncio.iscoroutinefunction` would support).

More importantly, AsyncMock's will evaluate to False in `inspect.iscoroutinefunction(AsyncMock)`, meaning setting the `side_effect` as an AsyncMock will return the coroutine obj instead of awaiting on the AsyncMock set as the `side_effect`. This functionality is debatable, but "chaining" calls like this might be surprising and confusing for users, vs returning the coroutine itself.

Reply:
I moved to using asyncio.iscoroutinefunction() to support treating AsyncMocks as coroutines,  but after 3.10 it should probably be `inspect.iscoroutinefunction(obj) or isinstance(obj, AsyncMock)`